File: Problems/0100_0199/0114_Flatten_Binary_Tree_to_Linked_List/Project_Python3/TreeNode/Codec.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from .TreeNode import TreeNode

class Codec:
    # 112ms

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        def helper(node):
            if node:
                vals.append(str(node.val))
                helper(node.left)
                helper(node.right)
            else:
                # vals.append('#')
                vals.append('null')
        vals = []
        helper(root)
        return ' '.join(vals)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        def helper():
            try:
                val = next(vals)
                if val == '#':
                    return None
                elif val == "null":
                    return None
                else:
                    node = TreeNode(int(val))
                    node.left = helper()
                    node.right = helper()
                return node
            except:
                return None
        vals = iter(data.split(","))
        return helper()

File: Problems/0100_0199/0110_Balanced_Binary_Tree/Project_Java/Codec.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

import java.util.*;

public class Codec {
    // 8ms

    private static final String SPLITTER = ",";
    private static final String NULLSTR = "null";

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        buildString(root, sb);
        int size = sb.length();
        return sb.delete(size - 1, size).toString();
    }

    private void buildString(TreeNode node, StringBuilder sb) {
        if (node == null) {
            sb.append(NULLSTR).append(SPLITTER);
        } else {
            sb.append(node.val).append(SPLITTER);
            buildString(node.left, sb);
            buildString(node.right,sb);
        }
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        Deque<String> nodes = new LinkedList<>();
        nodes.addAll(Arrays.asList(data.split(SPLITTER)));
        return buildTree(nodes);
    }

    private TreeNode buildTree(Deque<String> nodes) {
        if (nodes.size() <= 0) {
            return null;
        }

        String val = nodes.remove();
        if (val.equals(NULLSTR)) {
            return null;
        } else {
            TreeNode node;
            if (val.equals("null")) {
                node = null;
            } else {
                node = new TreeNode(Integer.valueOf(val));
                node.left = buildTree(nodes);
                node.right = buildTree(nodes);
            }
            return node;
        }
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));
